File: app/Http/Controllers/Api/V1/ProductApiController.php
<?php

namespace App\Http\Controllers\Api\V1;

use App\Http\Controllers\Controller;
use App\Services\ProductService;
use Illuminate\Http\Request;

class ProductApiController extends Controller
{
    protected $productService;

    public function __construct(ProductService $productService)
    {
        $this->productService = $productService;
    }

    public function index()
    {
        return $this->productService->getAllProducts();
    }

    public function store(Request $request)
    {
        return $this->productService->createProduct($request->all());
    }

    public function show($id)
    {
        return $this->productService->getProductById($id);
    }

    public function destroy($id)
    {
        return $this->productService->deleteProduct($id);
    }

    public function update(Request $request, $id)
    {
        return $this->productService->updateProduct($id, $request->all());
    }
}

File: spring-project/src/main/java/com/example/demo/ProductApiController.java
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/products")
public class ProductApiController {

    private final ProductService productService;

    @Autowired
    public ProductApiController(ProductService productService) {
        this.productService = productService;
    }

    @GetMapping
    public List<Product> index() {
        return productService.getAllProducts();
    }

    @PostMapping
    public Product store(@RequestBody ProductRequest request) {
        return productService.createProduct(request);
    }

    @GetMapping("/{id}")
    public Product show(@PathVariable("id") Long id) {
        return productService.getProductById(id);
    }

    @DeleteMapping("/{id}")
    public void destroy(@PathVariable("id") Long id) {
        productService.deleteProduct(id);
    }

    @PutMapping("/{id}")
    public Product update(@PathVariable("id") Long id, @RequestBody ProductRequest request) {
        return productService.updateProduct(id, request);
    }
}